src/librustc_mir/const_eval.rs

```diff
@@ -93,15 +72,36 @@ fn op_to_const<'tcx>(
         _ => false,
     };
     let normalized_op = if normalize {
-        Err(*ecx.read_immediate(op).expect("normalization works on validated constants"))
+        Err(ecx.read_immediate(op).expect("normalization works on validated constants"))
     } else {
+        // FIXME(oli-obk): if we have an `Immediate::ScalarPair` here, that seems bad as
+        // we only want to allow slices.
         op.try_as_mplace()
     };
     let val = match normalized_op {
-        Ok(mplace) => return mplace_to_const(ecx, mplace),
-        Err(Immediate::Scalar(x)) =>
+        Ok(mplace) => {
+            let ptr = mplace.ptr.to_ptr()?;
+            let alloc = ecx.tcx.alloc_map.lock().unwrap_memory(ptr.alloc_id);
+            if alloc.align == mplace.align {
```

Comment:
Looks like this could be an inequality test as well (`alloc.align >= mplace.align`, I'd thing)?

Reply:
We can add an assertion that `alloc.align >= mplace.align`, we should have errored out during const eval if that is violated.

Reply:
You mean, we can assert instead of copy? That'd be great!

Reply:
Any update on this?